django/thunderstore/api/cyberstorm/views/markdown.py

```diff
@@ -0,0 +1,77 @@
+from typing import Optional
+
+from django.http import Http404
+from rest_framework import serializers
+from rest_framework.generics import RetrieveAPIView, get_object_or_404
+
+from thunderstore.api.utils import CyberstormAutoSchemaMixin
+from thunderstore.markdown.templatetags.markdownify import render_markdown
+from thunderstore.repository.models import Package, PackageVersion
+
+
+class CyberstormMarkdownResponseSerializer(serializers.Serializer):
+    html = serializers.CharField()
+
+
+class PackageVersionReadmeAPIView(CyberstormAutoSchemaMixin, RetrieveAPIView):
+    """
+    Return README.md prerendered as HTML.
+
+    If no version number is provided, the latest version is used.
+    """
+
+    serializer_class = CyberstormMarkdownResponseSerializer
+
+    def get_object(self):
+        package_version = get_package_version(
+            self.kwargs["namespace_id"],
+            self.kwargs["package_name"],
+            self.kwargs.get("version_number"),
+        )
+
+        return {"html": render_markdown(package_version.readme)}
+
+
+class PackageVersionChangelogAPIView(CyberstormAutoSchemaMixin, RetrieveAPIView):
+    """
+    Return CHANGELOG.md prerendered as HTML.
+
+    If no version number is provided, the latest version is used.
+    """
+
+    serializer_class = CyberstormMarkdownResponseSerializer
+
+    def get_object(self):
+        package_version = get_package_version(
+            self.kwargs["namespace_id"],
+            self.kwargs["package_name"],
+            self.kwargs.get("version_number"),
+        )
+
+        if package_version.changelog:
+            return {"html": render_markdown(package_version.changelog)}
+
+        return {"html": ""}
+
+
+def get_package_version(
+    namespace_id: str,
+    package_name: str,
+    version_number: Optional[str],
+) -> PackageVersion:
+    package = get_object_or_404(
+        Package.objects.active(),
+        namespace__name=namespace_id,
+        name__iexact=package_name,
+    )
+
+    if version_number:
+        return get_object_or_404(
+            package.versions.active(),
+            version_number=version_number,
+        )
+
+    if package.latest.is_active:
+        return package.latest
+
+    raise Http404
```

Comment:
This method really confused me until I realized the views are intended to work with and without the version number.

Now I see why it's done the way it is, but I suspect we could still handle this with just 1 query by handling the conditionals before querying

Perhaps something along the lines of this:
```python
    filters = Q(
        is_active=True,
        package__is_active=True,
        package__namespace__name=namespace_id,
        name=package_name,
    )

    if version_number:
        filters &= Q(version_number=version_number)
    else:
        filters &= Q(package__latest=F("id"))

    return get_object_or_404(PackageVersion, filters)
```

Reply:
You seem easily confused. Noted the improvement in the task nevertheless.